# Eigenvalues and Eigenvectors of a Matrix

Inspired by:
https://math.libretexts.org/Bookshelves/Linear_Algebra/A_First_Course_in_Linear_Algebra_(Kuttler)/07:_Spectral_Theory/7.01:_Eigenvalues_and_Eigenvectors_of_a_Matrix

---

## Step 1 - What is the problem?

We start with a square matrix $ A $ and ask:

> Does there exist a non-zero vector whose direction is unchanged by $ A $?

Mathematically:

$
A v = \lambda v
$

We are looking for:
- $ v \neq 0 $
- direction is preserved
- only scaling changes

---

## Step 2 - Geometric meaning

A [matrix transformation](https://en.wikipedia.org/wiki/Transformation_matrix) usually:
- rotates vectors
- stretches them differently in different directions
- distorts geometry

But eigenvectors are **special directions** where the transformation behaves like pure scaling:

$
A v \parallel v
$

That means:
> after applying $A$, the vector still lies on the same line through the origin

In [ ]:
import numpy as np

A = np.array([
    [0, 5, -10],
    [0, 22, 16],
    [0, -9, -2],
])

X1 = np.array([-5, -4, 3])
X2 = np.array([1, 0, 0])

In [3]:
A @ X1, X1 * 10

(array([-50, -40,  30]), array([-50, -40,  30]))

In [4]:
A @ X2, X2 * 0

(array([0, 0, 0]), array([0, 0, 0]))

## Matrix–Vector Multiplication Example

We are given:

$$
A =
\begin{bmatrix}
0 & 5 & -10 \\
0 & 22 & 16 \\
0 & -9 & -2
\end{bmatrix}
\quad
X_1 =
\begin{bmatrix}
-5 \\
-4 \\
3
\end{bmatrix}
$$

We want to compute:

$$
A X_1
$$

---

**Step 1 — [Rule of matrix multiplication](https://en.wikipedia.org/wiki/Matrix_multiplication#Matrix_times_vector)**

Each entry of the result is computed as a **[dot product](https://en.wikipedia.org/wiki/Dot_product) between a row of $A$** and the vector $X_1$.

---

**Step 2 — First row**

$$
[0,\ 5,\ -10] \cdot [-5,\ -4,\ 3]
$$

$$
= 0 \cdot (-5) + 5 \cdot (-4) + (-10) \cdot 3
$$

$$
= 0 - 20 - 30 = -50
$$

---

**Step 3 — Second row**

$$
[0,\ 22,\ 16] \cdot [-5,\ -4,\ 3]
$$

$$
= 0 \cdot (-5) + 22 \cdot (-4) + 16 \cdot 3
$$

$$
= 0 - 88 + 48 = -40
$$

---

**Step 4 — Third row**

$$
[0,\ -9,\ -2] \cdot [-5,\ -4,\ 3]
$$

$$
= 0 \cdot (-5) + (-9) \cdot (-4) + (-2) \cdot 3
$$

$$
= 0 + 36 - 6 = 30
$$

---

**Final result**

$$
A X_1 =
\begin{bmatrix}
-50 \\
-40 \\
30
\end{bmatrix}
$$

---

**Key insight (preview of eigenvectors)**

Notice that:

$$
A X_1 = 10 \cdot X_1
$$

This means:
- the vector direction does not change
- only its length is scaled by 10

So $X_1$ behaves like an **eigenvector of $A$** with eigenvalue $ \lambda = 10 $.

## Step 3 - Algebraic Form

We start from the eigenvalue equation:

$
A v = \lambda v
$

The goal is to find vectors $v \neq 0$ whose direction does not change under the transformation by $A$.

To make this easier to analyze, move everything to one side:

$
A v - \lambda v = 0
$

Factor out $v$:

$
(A - \lambda I) v = 0
$

where $I$ is the [identity matrix](https://en.wikipedia.org/wiki/Identity_matrix).

This transforms the eigenvalue problem into a [**homogeneous linear system**](https://www.bohrium.com/en/sciencepedia/feynman/linear_algebra_undergraduate-homogeneous_linear_systems).

---

## Step 4 - When does a non-trivial solution exist?

The equation

$
(A - \lambda I) v = 0
$

always has the [**trivial solution**](https://www.cuemath.com/algebra/homogeneous-system-of-linear-equations/):

$
v = 0
$

But we are not interested in this.  
We want a **non-zero vector** $v \neq 0$.

From linear algebra, a homogeneous system has [non-trivial solutions](https://amixsystems.com/homogeneous-system/#solution-types). **only when the matrix is not [invertible](https://en.wikipedia.org/wiki/Invertible_matrix)**

[A matrix is not invertible when its determinant is zero](https://math.stackexchange.com/questions/2093067/why-do-homogenous-linear-systems-have-determinant-zero):

$
\det(A - \lambda I) = 0
$

This equation is called the [**characteristic equation**](https://en.wikipedia.org/wiki/Characteristic_polynomial).

Solving it gives the values of $ \lambda $ for which eigenvectors exist.

These values are called **eigenvalues**.

## Step 5 - Eigenvalues Come First (Characteristic Equation)

From the previous step, we know that a non-trivial solution to

$
(A - \lambda I)v = 0
$

exists only when

$
\det(A - \lambda I) = 0
$

This determinant produces a **polynomial in $ \lambda $** called the:

> [**characteristic polynomial**](https://en.wikipedia.org/wiki/Characteristic_polynomial)

Solving this equation gives the values of $ \lambda $ for which eigenvectors exist.

These values are called **eigenvalues** of the matrix $A$.

So the process always starts with:

$
\det(A - \lambda I) = 0
\quad \Longrightarrow \quad
\text{find eigenvalues}
$

---


## Step 6 - Eigenvectors Come Next

Once an eigenvalue $ \lambda $ is known, we substitute it back into:

$
(A - \lambda I)v = 0
$

This becomes a standard homogeneous linear system.

Solving this system gives the corresponding **eigenvectors**.

So the overall procedure is:

1. Compute $ \det(A - \lambda I) $
2. Solve for eigenvalues $ \lambda $
3. For each $ \lambda $, solve $ (A - \lambda I)v = 0 $ to find eigenvectors

## Step 7 - Key Interpretation

An eigenpair $ (\lambda, v) $ satisfies:

$
A v = \lambda v
$

This means:

- The vector $v$ keeps its **direction** after the transformation by $A$
- The scalar $ \lambda $ tells how much the vector is **stretched or shrunk**

So, along eigenvectors, the matrix transformation behaves like **pure scaling**.

In other words:

> A matrix can be understood by the special directions where it acts as simple multiplication.

## Finding Eigenvalues and Eigenvectors - Worked Example

Consider the matrix

$
A =
\begin{bmatrix}
-5 & 2 \\
-7 & 4
\end{bmatrix}
$

We will now follow the exact procedure developed earlier:

1. Form $ A - \lambda I $
2. Compute $ \det(A - \lambda I) $ — the characteristic equation
3. Solve for eigenvalues $ \lambda $
4. Substitute each eigenvalue into $ (A - \lambda I)v = 0 $ to find eigenvectors

In [5]:
# Step 1 — Define the matrix
A = np.array([
    [-5, 2],
    [-7, 4],
])

In [6]:
# Step 2 — Compute A−λI symbolically
import sympy as sp

lam = sp.symbols('lam')

A_sym = sp.Matrix([
    [-5, 2],
    [-7, 4]
])

I = sp.eye(2)

A_minus_lamI = A_sym - lam * I
A_minus_lamI

Matrix([
[-lam - 5,       2],
[      -7, 4 - lam]])

In [7]:
# Step 3 — Characteristic equation det(A−λI)=0
char_poly = A_minus_lamI.det()
char_poly

lam**2 + lam - 6

In [ ]:
# Step 4 — Solve for eigenvalues λ^2+λ-6=0
eigenvalues = sp.solve(char_poly, lam)
eigenvalues

[-3, 2]

In [ ]:
# Step 5 — Compute eigenvectors as (λI - A) X = 0 where X is eigenvector
# (-3I - A) X = 0 - first 
# (2I - A) X = 0 - second 
eigenvectors = A_sym.eigenvects()
eigenvectors

[(-3,
  1,
  [Matrix([
   [1],
   [1]])]),
 (2,
  1,
  [Matrix([
   [2/7],
   [  1]])])]

**Clean summary**
| eigenvalue | multiplicity | eigenvector |
| ---------- | ------------ | ----------- |
| -3         | 1            | ([1, 1])    |
| 2          | 1            | ([2/7, 1])  |

second vector could be multiplied by 7 and get [2, 7] - still a perfectly valid eigenvector

In [10]:
# Step 6 — Verify using NumPy (numeric check)
eigvals_np, eigvecs_np = np.linalg.eig(A)
eigvals_np, eigvecs_np

(array([-3.,  2.]),
 array([[-0.70710678, -0.27472113],
        [-0.70710678, -0.96152395]]))

[-0.70710678, -0.70710678] the same as [1, 1]

[-0.27472113, -0.96152395] the same as [2/7, 1]

-0.2747 / -0.9615 ≈ 0.2857 ≈ 2 / 7

In [13]:
for i in range(len(eigvals_np)):
    v = eigvecs_np[:, i]
    lam = eigvals_np[i]

    lhs = A @ v
    rhs = lam * v

    print(f"\nEigenvalue {lam}")
    print("A v =", lhs)
    print("λ v =", rhs)
    print("close?", np.allclose(lhs, rhs))


Eigenvalue -3.0
A v = [2.12132034 2.12132034]
λ v = [2.12132034 2.12132034]
close? True

Eigenvalue 2.0
A v = [-0.54944226 -1.9230479 ]
λ v = [-0.54944226 -1.9230479 ]
close? True


### Important Note - Eigenvector Scaling

Eigenvectors are **not unique vectors**, they are **directions**.

If $v$ is an eigenvector of $A$, then any scaled version is also an eigenvector:

$
A v = \lambda v \quad \Rightarrow \quad A (c v) = \lambda (c v)
$

for any non-zero scalar $c$.

---

**Example**

These vectors represent the **same eigenvector direction**:

$
\begin{bmatrix} 1 \\ 1 \end{bmatrix}
\quad \text{and} \quad
\begin{bmatrix} -0.707 \\ -0.707 \end{bmatrix}
$

They differ only by scaling, but define the same geometric object:

> a line (1D subspace) through the origin.

---

**What changes and what does not**

| Representation | Meaning |
|------|--------|
| \([1, 1]\) | Eigenvector direction (not normalized) |
| \([-0.707, -0.707]\) | Same direction, unit-length version |
| both | Same geometric line (same eigenspace) |

---

**What PCA uses**

PCA uses **unit eigenvectors** (orthonormal basis):

- ensures clean projections
- makes variance interpretation valid
- removes scaling ambiguity

So in PCA:

> only direction matters, not magnitude
```

### 3 x 3 matrix

In [15]:
# Step 1 — Define the matrix
A = np.array([
    [5, 10, -5],
    [2, 14, 2],
    [-4, -8, 6],
])

# Step 2 — Compute A−λI symbolically
lam = sp.symbols('lam')

A_sym = sp.Matrix([
    [5, 10, -5],
    [2, 14, 2],
    [-4, -8, 6],
])

I = sp.eye(3)

A_minus_lamI = A_sym - lam * I
A_minus_lamI

Matrix([
[5 - lam,       10,      -5],
[      2, 14 - lam,       2],
[     -4,       -8, 6 - lam]])

In [16]:
# Step 3 — Characteristic equation det(A−λI)=0
char_poly = A_minus_lamI.det()
char_poly

-lam**3 + 25*lam**2 - 160*lam + 100

In [17]:
# Step 4 — Solve for eigenvalues -λ^3+25λ^2-160λ+100=0
eigenvalues = sp.solve(char_poly, lam)
eigenvalues

[10, 15/2 - sqrt(185)/2, sqrt(185)/2 + 15/2]

In [18]:
# Step 5 — Compute eigenvectors as (λI - A) X = 0 where X is eigenvector
# (10I - A) X = 0 - first 
eigenvectors = A_sym.eigenvects()
eigenvectors

[(10,
  1,
  [Matrix([
   [-1],
   [ 0],
   [ 1]])]),
 (15/2 - sqrt(185)/2,
  1,
  [Matrix([
   [5/8 + sqrt(185)/8],
   [             -1/2],
   [                1]])]),
 (sqrt(185)/2 + 15/2,
  1,
  [Matrix([
   [5/8 - sqrt(185)/8],
   [             -1/2],
   [                1]])])]

In [19]:
# Step 6 — Verify using NumPy (numeric check)
eigvals_np, eigvecs_np = np.linalg.eig(A)
eigvals_np, eigvecs_np

(array([ 0.69926475, 14.30073525, 10.        ]),
 array([[ 9.01228636e-01, -6.93159341e-01, -7.07106781e-01],
        [-1.93797288e-01, -3.22344576e-01,  7.02166694e-17],
        [ 3.87594577e-01,  6.44689152e-01,  7.07106781e-01]]))

In [21]:
(15-np.sqrt(185))/2, (15+np.sqrt(185))/2

(np.float64(0.699264745632278), np.float64(14.300735254367723))

#### Checking Eigenvalues and Eigenvectors (SymPy vs NumPy)

We computed eigenvalues and eigenvectors using both **SymPy (exact)** and **NumPy (numerical)**.

##### Eigenvalues comparison

**SymPy (exact form)**

$
\lambda_1 = 10,\quad
\lambda_2 = \frac{15 - \sqrt{185}}{2},\quad
\lambda_3 = \frac{15 + \sqrt{185}}{2}
$

**NumPy (numerical approximation)**

```text
[0.69926475, 14.30073525, 10.0]
```

**Verification**

Approximating SymPy values:

- $ \frac{15 - \sqrt{185}}{2} \approx 0.699 $
- $ \frac{15 + \sqrt{185}}{2} \approx 14.301 $

✔ These match the NumPy results exactly (up to numerical precision).

---

##### Eigenvector comparison

**SymPy eigenvector for $ \lambda = 10 $**

$
v =
\begin{bmatrix}
-1 \\
0 \\
1
\end{bmatrix}
$

**NumPy eigenvector (approx.)**

$
v =
\begin{bmatrix}
-0.707 \\
0 \\
0.707
\end{bmatrix}
$

## A zero eigenvalue?

Given:

$
A =
\begin{bmatrix}
2 & 2 & -2 \\
1 & 3 & -1 \\
-1 & 1 & 1
\end{bmatrix}
$

---

**1. Key idea**

A matrix has a **zero eigenvalue** if:

$
\det(A) = 0
$

because this implies that $ \lambda = 0 $ is a root of:

$
\det(A - \lambda I) = 0
$

---

**2. How to check**

We can check this in two equivalent ways:

**(a) [Determinant](https://en.wikipedia.org/wiki/Determinant) test**

If:

$
\det(A) = 0
$

then the matrix has a zero eigenvalue.

**(b) Eigenvalue computation**

Compute:

$
\text{eigvals}(A)
$

and check whether 0 (or a very small value close to 0) appears.

---

**3. Geometric interpretation**

If $ \lambda = 0 $ is an eigenvalue, then there exists a non-zero vector $v$ such that:

$
A v = 0
$

This means:

- the matrix **collapses a direction to zero**
- the transformation is **not invertible**
- the matrix loses one effective dimension

---

**4. Impact on PCA**

In PCA (applied to a covariance matrix):

- eigenvalues represent **variance along principal directions**
- eigenvectors represent **principal directions**

If $ \lambda = 0 $:

- there is **no variance in that direction**
- the data lies in a **lower-dimensional subspace**
- that principal component carries **no information**

So in PCA:

> a zero eigenvalue means the corresponding principal component can be completely discarded without losing information

---

**5. Summary rule**

| Condition | Meaning |
|------|--------|
| $ \det(A) = 0 $ | zero eigenvalue exists |
| $ \det(A) \neq 0 $ | no zero eigenvalue |
| <a href="https://en.wikipedia.org/wiki/Rank_(linear_algebra)">rank</a> < n | zero eigenvalue guaranteed |

---

**6. Intuition**

A zero eigenvalue means:

> there is a direction in space that gets completely flattened to the origin by the transformation

In [25]:
A = np.array([
    [2, 2, -2],
    [1, 3, -1],
    [-1, 1, 1],
])

np.linalg.eigvals(A).round(0)

array([ 4., -0.,  2.])

Find same and other examples including Eigenvalues and Eigenvectors for Special Types of Matrices on [Eigenvalues and Eigenvectors of a Matrix](https://math.libretexts.org/Bookshelves/Linear_Algebra/A_First_Course_in_Linear_Algebra_%28Kuttler%29/07%3A_Spectral_Theory/7.01%3A_Eigenvalues_and_Eigenvectors_of_a_Matrix) 